pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            ensure!(AreaData::<T>::contains_key(root_id, area_id), Error::<T>::NotExists);
+            AreaData::<T>::mutate(root_id, area_id, |ar| {
+                ar.area_type = area_type;
+            });
+            Self::deposit_event(RawEvent::AreaTypeChanged(area_type, area_id, root_id, who));
+            Ok(())
+        }
+    }
+}
+
+// Module allows  use  common functionality by dispatchables
+impl<T: Trait> Module<T> {
+    // Implement module function.
+    // Public functions can be called from other runtime modules.
+    // TODO implement safe math
+    fn detect_touch(root_box: RootBoxOf<T>, touch: Point2D<T::Coord>) -> T::AreaId {
+        let root_base_point: Point2D<T::Coord> = 
+        Point2D::new(root_box.bounding_box.north_west.lat,
+                     root_box.bounding_box.north_west.lon); 
+        let root_secondary_point: Point2D<T::Coord> = 
+        Point2D::new(root_box.bounding_box.south_east.lat,
+                    root_box.bounding_box.south_east.lon); 
+        let root_dimensions = Self::get_distance_vector(root_base_point, root_secondary_point);
+        let distance_vector = Self::get_distance_vector(root_base_point, touch);
+        // TODO handle possible overflow errors
+        let row: u16 = (distance_vector.lat / root_box.delta).into().to_num::<u16>() + 1;
+        let column: u16 = (distance_vector.lon / root_box.delta).into().to_num::<u16>() + 1;
+        let total_rows: u16 = (root_dimensions.lat / root_box.delta).into().to_num::<u16>();
+
+        ((total_rows * (column - 1)) + row).into()
+    }
+
+    fn get_distance_vector( first_point: Point2D<T::Coord>, 
+                            second_point: Point2D<T::Coord>) -> Point2D<T::Coord> {
+        // code is clumsy, guess may be simplified
+        let lat_length: T::Coord;
+        if first_point.lat > second_point.lat {
+            lat_length = first_point.lat - second_point.lat;
+        } else {
+            lat_length = second_point.lat - first_point.lat;
+        }
+        let long_length: T::Coord; 
+        if first_point.lon > second_point.lon {
+            long_length = first_point.lon - second_point.lon; 
+        } else {
+            long_length = second_point.lon - first_point.lon;
+        }
+
+        Point2D::new(lat_length, long_length)
+    }
+    
+    /// form index for storing zones, wrapped in u64
+    // v.............root id here............v v.....area id.....v v..child objects..v
+    // 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000
+    fn form_index(root: T::RootId, area: T::AreaId, childs: u16) -> T::ZoneId {
+        let root: u32 = root.into();
+        let area: u16 = area.into();
+
+        ((root as u64) << 32 |
+         (area as u64) << 16 | 
+         childs as u64).into()      
```

Comment:
All this is fishy. If root/area/child id must be of specific size (u64/u32/u16) and it wouldn't work with any other data types, why at all to parameterize it and make an illusion of possibility to use different types?